semantics-3.0/Semantics.hs

```diff
@@ -236,206 +227,219 @@ data ReduceWarning = ReduceNoWarning
-        (input : rest) -> case apply env curState input cont of
-            Applied newState cont ->
-                applyAllAux env newState cont rest (warnings ++ warns) (effects ++ effs)
-            ApplyError error -> AAApplyError error
-
-
-applyAll :: Environment -> State -> Contract -> [Input] -> ApplyAllResult
-applyAll env state contract inputs = applyAllAux env state contract inputs [] []
+data ApplyAllResult = ApplyAllSuccess [ReduceWarning] [Payment] State Contract
+                    | ApplyAllNoMatchError
+                    | ApplyAllAmbiguousSlotIntervalError
+  deriving (Eq,Ord,Show)
+
+
+-- | Apply a list of Inputs to the contract
+applyAllInputs :: Environment -> State -> Contract -> [Input] -> ApplyAllResult
+applyAllInputs env state contract inputs = let
+    applyAllAux
+        :: Environment
+        -> State
+        -> Contract
+        -> [Input]
+        -> [ReduceWarning]
+        -> [Payment]
+        -> ApplyAllResult
+    applyAllAux env state contract inputs warnings effects =
+        case reduceContractUntilQuiescent env state contract of
+            ReduceAllAmbiguousSlotIntervalError -> ApplyAllAmbiguousSlotIntervalError
+            ReduceAllSuccess warns effs curState cont -> case inputs of
+                [] -> ApplyAllSuccess (warnings ++ warns) (effects ++ effs) curState cont
+                (input : rest) -> case applyInput env curState input cont of
+                    Applied newState cont ->
+                        applyAllAux env newState cont rest (warnings ++ warns) (effects ++ effs)
+                    ApplyNoMatchError -> ApplyAllNoMatchError
+    in applyAllAux env state contract inputs [] []
 
 -- PROCESS
 
--- List of signatures needed by a transaction
-type TransactionSignatures = Set Party
-
-data ProcessError = PEReduceError ReduceError
-                  | PEApplyError ApplyError
+data ProcessError = PEAmbiguousSlotIntervalError
+                  | PEApplyNoMatchError
                   | PEIntervalError IntervalError
                   | PEUselessTransaction
-  deriving (Eq,Ord,Show,Read)
+  deriving (Eq,Ord,Show)
 
-type ProcessWarning = ReduceWarning
-type ProcessEffect = ReduceEffect
 
-data ProcessResult = Processed [ProcessWarning]
-                               [ProcessEffect]
-                               TransactionSignatures
+data ProcessResult = Processed [ReduceWarning]
+                               [Payment]
                                TransactionOutcomes
                                State
                                Contract
                    | ProcessError ProcessError
-  deriving (Eq,Ord,Show,Read)
+  deriving (Eq,Ord,Show)
 
 data Transaction = Transaction { txInterval :: SlotInterval
                                , txInputs   :: [Input] }
-  deriving (Eq,Ord,Show,Read)
+  deriving (Eq,Ord,Show)
+
+type TransactionOutcomes = Map Party Money
+
 
+emptyOutcome :: TransactionOutcomes
+emptyOutcome = Map.empty
```

Comment:
why not just use `mempty` wherever you need (just in the fold I think).

Reply:
I guess, to not introduce another dependency on Monoid typeclass, and to self-document the code.

Reply:
is there an issue introducing dependencies?

Reply:
It's not an issue. Rather, it's a question of necessity. I'd inline `emptyOutcome` into `getOutcomes`